### PR TITLE
ovl_Obj_Demo decompiled

### PR DESCRIPTION
**Pull request by sonicdcer**

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 1]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
+            this->unk_150 = temp_v0 * 4.0f;
+        }
+        func_80983634(globalCtx);
+        this->actionFunc = func_80983678;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Demo/func_80983634.s")
+void func_80983634(GlobalContext* globalCtx) {
+    if ((globalCtx->sceneNum == 0x1D) && (func_801A8A50(0) == 0x6F)) {
```
Comment (tom-overton) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:65]:
```suggestion
    if ((globalCtx->sceneNum == SCENE_CASTLE) && (func_801A8A50(0) == NA_BGM_IKANA_CASTLE)) {
```

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 2]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
+        }
+        if (thisx->shape.rot.x == 0) {
+            this->unk_14C = 80.0f;
+        } else {
+            this->unk_14C = thisx->shape.rot.x * 4.0f;
+            thisx->shape.rot.x = 0;
+            thisx->world.rot.x = 0;
+        }
+        temp_v0 = thisx->shape.rot.z;
+        if (temp_v0 == 0) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:52]: https://decomp.me/scratch/Rr4QY can get rid of the temp by moving the assignment up so it uses a different register. Matches the x case further up better too.

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 3]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
+void func_80983634(GlobalContext* globalCtx) {
+    if ((globalCtx->sceneNum == SCENE_CASTLE) && (func_801A8A50(0) == NA_BGM_IKANA_CASTLE)) {
+        Audio_QueueSeqCmd(0x100100FF);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Demo/func_80983678.s")
+void func_80983678(ObjDemo* this, GlobalContext* globalCtx) {
+    func_80983634(globalCtx);
+    if ((this->actor.xzDistToPlayer < this->unk_14C) && (fabsf(this->actor.playerHeightRel) < this->unk_150)) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:72]: This tells us what 14C and 150 are: xzRange and yRange.

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 4]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Demo/func_80983704.s")
+void func_80983704(ObjDemo* this, GlobalContext* globalCtx) {
+    if ((this->unk_148 == 1) && (ActorCutscene_GetCurrentIndex() == 0x7D)) {
+        ActorCutscene_Stop(0x7D);
+        ActorCutscene_SetIntentToPlay(this->actor.cutscene);
+        return;
+    }
+    if (ActorCutscene_GetCanPlayNext(this->actor.cutscene) != 0) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:87]:
```suggestion
    if (ActorCutscene_GetCanPlayNext(this->actor.cutscene)) {
```

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 5]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
+        ActorCutscene_SetIntentToPlay(this->actor.cutscene);
+        func_80983634(globalCtx);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Demo/ObjDemo_Update.s")
+void ObjDemo_Update(Actor* thisx, GlobalContext* globalCtx) {
+    ObjDemo* this = THIS;
+
+    if ((this->actor.params != 0xFF) && (Flags_GetSwitch(globalCtx, this->actor.params) != 0)) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:113]:
```suggestion
    if ((this->actor.params != 0xFF) && Flags_GetSwitch(globalCtx, this->actor.params)) {
```

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 6]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
     (ActorFunc)NULL,
 };
 
-#endif
+void ObjDemo_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ObjDemo* this = THIS;
+    s16 temp_v0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Demo/ObjDemo_Init.s")
+    thisx->params &= 0xFF;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:36]: This ought to be done with a GET params macro, and the special case 0xFF should have a define, at least until we find out what this is.

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 7]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
+    if ((thisx->params != 0xFF) && (Flags_GetSwitch(globalCtx, thisx->params))) {
+        Actor_MarkForDeath(&this->actor);
+    } else {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:39]:
```suggestion
        Actor_MarkForDeath(&this->actor);
        return;
    }
```
One of the few cases where this is preferred.

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c [hunk 8]
```diff
@@ -29,14 +29,90 @@ const ActorInit Obj_Demo_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Demo/ObjDemo_Update.s")
+void ObjDemo_Update(Actor* thisx, GlobalContext* globalCtx) {
+    ObjDemo* this = THIS;
+
+    if ((this->actor.params != 0xFF) && (Flags_GetSwitch(globalCtx, this->actor.params) != 0)) {
+        Actor_MarkForDeath(&this->actor);
+    } else {
+        this->actionFunc(this, globalCtx);
+    }
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:117]:
```suggestion
        Actor_MarkForDeath(&this->actor);
        return;
    }

    this->actionFunc(this, globalCtx);
```

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c (hunk 4 above)
Comment (hensldm) [reply src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:87]: Can also do an `else if` to remove the early `return`

**Review by hensldm — COMMENTED** [review]
Also need to merge and resolve conflicts.

File: src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c (hunk 6 above)
Comment (sonicdcer) [reply src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:36]: two macros? please specify for each

Comment (hensldm) [reply src/overlays/actors/ovl_Obj_Demo/z_obj_demo.c:35]:
```suggestion
    thisx->params = OBJDEMO_GET_FF(thisx);
```
Also a `#define OBJDEMO_FF_FF 0xFF` for the comparision one line below.

**Review by EllipticEllipsis — APPROVED** [review]

**Review by hensldm — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**